contracts/Presale.sol

```diff
@@ -0,0 +1,227 @@
+    }
+
+    function usdToTokens(uint256 roundIndex, uint256 amount) public view returns (uint256) {
+        RoundConfig memory _round = $rounds[roundIndex];
+        return amount * _round.tokenPrice;
+    }
+
+    function setConfig(PresaleConfig calldata newConfig) external override onlyOwner {
+        require(newConfig.startDate < newConfig.endDate, "INVALID_DATES");
+        require(newConfig.withdrawTo != address(0), "INVALID_WITHDRAW_TO");
+
+        emit ConfigUpdated($config, newConfig, _msgSender());
+
+        $config = newConfig;
+    }
+
+    function setRounds(RoundConfig[] calldata newRounds) external override onlyOwner {
+        emit RoundsUpdated($rounds, newRounds, _msgSender());
+
+        uint256 _totalRaisedUSD = $totalRaisedUSD;
+
+        uint256 totalCostUSD;
+        uint256 expectedCurrentRoundIndex;
+
+        for (uint256 i; i < newRounds.length; ++i) {
+            $rounds.push(newRounds[i]);
+
+            uint256 roundCostUSD = newRounds[i].tokensAllocated * newRounds[i].tokenPrice;
+            totalCostUSD += roundCostUSD;
+            if (_totalRaisedUSD > totalCostUSD) {
+                expectedCurrentRoundIndex++;
+            }
+        }
+
+        $currentRoundIndex = expectedCurrentRoundIndex;
+    }
+
+    function purchase() public payable override whenNotPaused {
+        uint256 amountUSD = ethToUsd(msg.value);
+        _sync($currentRoundIndex, address(0), amountUSD, _msgSender());
+    }
+
+    function purchase(address account) public payable override whenNotPaused {
+        uint256 amountUSD = ethToUsd(msg.value);
+        _sync($currentRoundIndex, address(0), amountUSD, account);
+    }
+
+    function purchaseUSDC(uint256 amount) external override whenNotPaused {
+        address sender = _msgSender();
+
+        IERC20(USDC).transferFrom(sender, address(this), amount);
+
+        uint256 amountScaled = amount * USDC_TO_WEI_PRECISION;
+        _sync($currentRoundIndex, USDC, amountScaled, sender);
+    }
+
+    function purchaseDAI(uint256 amount) external override whenNotPaused {
+        address sender = _msgSender();
+
+        IERC20(DAI).transferFrom(sender, address(this), amount);
+        _sync($currentRoundIndex, DAI, amount, sender);
+    }
+
+    receive() external payable {
+        purchase();
+    }
+
+    function _sync(uint256 roundIndex, address asset, uint256 amountUSD, address account) private {
+        PresaleConfig memory _config = $config;
+        RoundConfig[] memory _rounds = $rounds;
+
+        uint256 _userTokensAllocated = $userTokensAllocated[account];
+        uint256 userAllocationRemaining = _config.maxUserAllocation - _userTokensAllocated;
+
+        uint256 remainingUSD = amountUSD;
+        uint256 roundAllocationRemaining;
+
+        for (uint256 i = roundIndex; i < _rounds.length; ++i) {
+            RoundConfig memory _round = _rounds[i];
+            roundAllocationRemaining = _round.tokensAllocated - ($raisedUSD[i] * _round.tokenPrice);
```

Comment:
This value could be greater than the config, if the config changes. So we want to update this like so
```suggestion
            uint256 _currentlyAllocated = $roundAllocated[i];
            roundAllocationRemaining = _currentlyAllocated < _round.tokensAllocated ? _round.tokensAllocated - _currentlyAllocated : 0;
```